exchanges/okx/ws_requests.go

```diff
@@ -0,0 +1,357 @@
+}
+
+// WsCancelSpreadOrder cancels an incomplete spread order through the websocket connection.
+func (ok *Okx) WsCancelSpreadOrder(ctx context.Context, orderID, clientOrderID string) (*SpreadOrderResponse, error) {
+	if orderID == "" && clientOrderID == "" {
+		return nil, order.ErrOrderIDNotSet
+	}
+
+	arg := make(map[string]string)
+	if orderID != "" {
+		arg["ordId"] = orderID
+	}
+	if clientOrderID != "" {
+		arg["clOrdId"] = clientOrderID
+	}
+
+	id := strconv.FormatInt(ok.Websocket.AuthConn.GenerateMessageID(false), 10)
+
+	var resp []*SpreadOrderResponse
+	if err := ok.SendAuthenticatedWebsocketRequest(ctx, cancelSpreadOrderEPL, id, "sprd-cancel-order", []map[string]string{arg}, &resp); err != nil {
+		return nil, mergeErrors(err, resp)
+	}
+
+	return singleItem(resp)
+}
+
+// WSCancelAllSpreadOrders cancels all spread orders and return success message through the websocket channel.
+func (ok *Okx) WSCancelAllSpreadOrders(ctx context.Context, spreadID string) error {
+	arg := make(map[string]string, 1)
+	if spreadID != "" {
+		arg["sprdId"] = spreadID
+	}
+
+	id := strconv.FormatInt(ok.Websocket.AuthConn.GenerateMessageID(false), 10)
+
+	var resps []*ResponseResult
+	if err := ok.SendAuthenticatedWebsocketRequest(ctx, cancelAllSpreadOrderEPL, id, "sprd-mass-cancel", []map[string]string{arg}, &resps); err != nil {
+		return err
+	}
+
+	resp, err := singleItem(resps)
+	if err != nil {
+		return err
+	}
+
+	if !resp.Result {
+		return errCancelAllSpreadOrdersFailed
+	}
+
+	return nil
+}
+
+// SendAuthenticatedWebsocketRequest sends a websocket request to the server
+func (ok *Okx) SendAuthenticatedWebsocketRequest(ctx context.Context, epl request.EndpointLimit, id, operation string, payload, result any) error {
+	if operation == "" || payload == nil {
+		return errInvalidWebsocketRequest
+	}
+
+	outbound := &struct {
+		ID        string `json:"id"`
+		Operation string `json:"op"`
+		Arguments any    `json:"args"`
+		// TODO: Add ExpTime to the struct, the struct should look like this:
+		// ExpTime  string `json:"expTime,omitempty"` so a deadline can be set
+		// Request effective deadline. Unix timestamp format in milliseconds, e.g. 1597026383085
+	}{
+		ID:        id,
+		Operation: operation,
+		Arguments: payload,
+	}
+
+	incoming, err := ok.Websocket.AuthConn.SendMessageReturnResponse(ctx, epl, id, outbound)
+	if err != nil {
+		return err
+	}
+
+	intermediary := struct {
+		ID        string `json:"id"`
+		Operation string `json:"op"`
+		Code      int64  `json:"code,string"`
+		Message   string `json:"msg"`
+		Data      any    `json:"data"`
+		InTime    string `json:"inTime"`
+		OutTime   string `json:"outTime"`
+	}{
+		Data: result,
+	}
+
+	if err := json.Unmarshal(incoming, &intermediary); err != nil {
+		return err
+	}
+
+	switch intermediary.Code {
+	case 0:
+		return nil
+	case 1:
+		return errOperationFailed
+	case 2:
+		return errPartialSuccess
+	default:
+		return getStatusError(intermediary.Code, intermediary.Message)
```

Comment:
I think the main thing that I'd like to change from gk's request is the following:
```go

func parseErrors(result any, err error) error {
	rv := reflect.ValueOf(result)
	if rv.Kind() == reflect.Pointer && rv.Elem().Kind() == reflect.Slice {
		s := rv.Elem()
		for i := range s.Len() {
			v := s.Index(i)
			if v.Kind() == reflect.Interface {
				v = v.Elem()

			}
			if subErr, ok := v.Interface().(hasError); ok && subErr.Error() != nil {
				err = common.AppendError(err, fmt.Errorf("%s[%d]: %w", v.Type(), i+1, subErr.Error()))
			}
		}
	}
	return err
}
```
extracting it into its own function allows testing of the logic there instead of requiring an auth connection and request. See test comment